edx/xblock-lti-consumer
# [BD-24][TNL-7607] BB-3072: Partial Deep Linking Implementation

This PR adds the initial base code that allows performing LTI Deep Linking configuration launches.

This PR includes:
- Deep Linking related models.
- Support for Deep Link Launches (configuration) from LTI XBlock.
- Simplification of LTI launch flows, reducing the required number of launch steps.

_TODO:_
- [x] Full test coverage
- [x] Add missing token validation
- [x] Split into smaller diffs

**Break down in smaller PRs:**

I'm working on reducing the PR size by breaking this one into multiple smaller PRs. They need to be merged in the following order to avoid issues: 
1. Launch simplification PR: https://github.com/edx/xblock-lti-consumer/pull/125
2. Deep Linking Launch implementation: https://github.com/edx/xblock-lti-consumer/pull/126
3. Deep Linking Response endpoint implementation and `ltiResourceLink` support: https://github.com/edx/xblock-lti-consumer/pull/127

**Testing instructions:**
1. Check out this branch.
2. Apply migrations.
3. Enable `LTI_1P3_ENABLED` and `LTI_DEEP_LINKING_ENABLED` on `studio.yml` *FEATURES* dict.
4. Set up an LTI 1.3 tool and make sure to also copy over the `Tool Deep Link URL` from IMS's reference implementation.
5. Check that the normal LTI launch works.
6. Since there's no way of doing deep linking launches yet, apply a small patch to test the deep link launch:
```
diff --git a/lti_consumer/lti_xblock.py b/lti_consumer/lti_xblock.py
index df0cf16..3e5cadd 100644
--- a/lti_consumer/lti_xblock.py
+++ b/lti_consumer/lti_xblock.py
@@ -1378,7 +1378,7 @@ class LtiConsumerXBlock(StudioEditableXBlockMixin, XBlock):
             # Retrieve and set LTI 1.3 Launch start URL
             lti_block_launch_handler = get_lti_1p3_launch_start_url(
                 block=self,
-                deep_link_launch=False,
+                deep_link_launch=True,
                 hint=str(self.location)  # pylint: disable=no-member
             )
```
7. Perform the LTI Deep Link Launch, it should work the same way as the other launch.

**Reviewers:**

**Notes:**
I'm still working on this, and the diff size is likely to increase, should I start breaking this down into smaller PRs?
This is NOT a full deep linking implementation.
